src/nvim/lua/vim.lua

```diff
@@ -478,4 +478,105 @@ function vim.defer_fn(fn, timeout)
   return timer
 end
 
+--- Parse a version string into a table.
+---
+--@param ver (string): version string
+--@param strict (optional bool): if true, a semver string is expected;
+--       otherwise, just dotted numbers, possibly with junk at the end
+--@return table with keys major, minor, patch, pre and build if strict,
+--        otherwise with numerical indices
+function vim.parse_version(ver, strict)
+  local err_msg = string.format('invalid version string: "%s"', ver)
+
+  local numbers = {}
+  local cursor, end_number, number
+  repeat
+    _, end_number, number = string.find(ver, '^(%d+)%.?', cursor)
+    if number then
+      cursor = end_number + 1
+      table.insert(numbers, tonumber(number))
+    end
+  until not number
+
+  if #numbers == 0 or (strict and #numbers > 3) then
+    error(err_msg)
+  elseif not strict then
+    return numbers
```

Comment:
I don't like that sometimes this function returns a nicely formatted, informative map of semver versions and other times just returns a list of numbers.

I think we should just remove "non-strict" mode from this. If you just want to compare dotted numbers, it's fine to write the 4 lines of code you have above splitting the numbers. (Do we use non-strict mode anywhere internally)?

Reply:
I initially didn't have a non-strict mode, because I thought I'd just be replacing ad hoc string or numeric conversion based comparisons of Python versions where I had a good idea of the version string format. But then I came across the script-local function `s:version_cmp` in `health/provider`, which is used by some of the other provider checks and which just splits the version strings on periods, leniently converts the elements to numbers and compares them.

I concluded `s:version_cmp` should also be replaced by the new version comparison function, but suddenly I wasn't so sure about the range of inputs these additional provider checks would feed it. `s:version_cmp` can basically soldier through any kind of input, even if you pass in `'foo'` and `'bar'`, it won't give up. By contrast, the strict version of the new `cmp_versions` function doesn't allow more than three dotted numbers, because that's what semver syntax stipulates.

So I realized a couple of things:

1. Swapping out `s:version_cmp` for a strict `cmp_versions` might break some of the health checks, should they happen to involve version strings which don't conform to semver syntax (e.g. because they have some trailing junk, or because they have more than three numbers).
2. Environment detection and troubleshooting is hard to test because it involves external moving parts, so maybe there's value in having a more resilient mode for that.
3. In strict semver syntax mode, there's just no way to parse version strings consisting of more than three dotted numbers. Still, being able to do this can occasionally be useful, even within semver itself -- the pre-release label is fairly freeform, it can consist of any number of dotted numbers, among other things, so when `cmp_versions` reaches the point where it needs to call itself recursively to compare pre-release labels, it does so in non-strict mode for precisely this reason.

So I implemented a non-strict mode, which is still stricter than `s:version_cmp` because it at least expects some numbers with periods between them, which seems like a reasonable baseline (so it would choke on `'foo'` and `'bar'` as invalid version strings). I made it the default, but it could conceivably be made opt-in instead if that would help you warm up to the notion :)

If none of this sounds convincing -- it's entirely possible I'm being too cautious -- then I guess let's remove it. I'll try and check the format of the version strings returned by those other provider programs which are currently compared via `s:version_cmp`, make any adjustments necessary so that the comparisons in the health checks don't error out even under strict `cmp_versions`, and we'll hope it doesn't break in the future or if people use weird pre-release versions etc.